Added functionality to to ensure all output is printed on new lines.

Andy53 opened the pull request:
Added a check to SemanticTest.cpp to ensure all output is printed on new lines. Also added a test (emit_multiple_outputs.sol). Fixes #13592

axic committed:
EVMHost: Simplify some code and dcoument functions

axic committed:
EVMHost: Calculate address for CREATE properly

axic committed:
Update tests and include new for create/create2 calculation

christianparpart committed:
Renaming lspAssert to lspRequire plus adding some documentation to make it more clear what this function is supposed to be used for.

gnattishness committed:
Add v0.6.0 change notes for inline assembly opcode syntax changes.

miles170 committed:
Fix stripping in CLI tests does not produce correct JSON formatting

miles170 committed:
Pretty print expected json output of command line tests

nikola-matic committed:
Add code style rule for std:: qualifier

0xGeorgii committed:
Update CVC4::BitVector ctor call

axic committed:
EVMHost: Make it clear which MockedHost helpers are used

And also remove unused helpers.

axic committed:
EVMHost: Reduce code duplication in precompiles

axic committed:
EVMHost: Add size%192 check to bn128pairing

And also refactor resultWithFailure

axic committed:
EVMHost: replace precompileALTBN128PairingProduct with mapping

axic committed:
Replace use of boost::adapters::filtered with ranges::views::filter

axic committed:
Remove unused include of boost/variant

axic committed:
Replace use of boost::algorithm::all_of with ranges::all_of

franzihei committed:
Update meeting link to meet.soliditylang.org

axic committed:
Improve FunctionSelector helpers

axic committed:
Drop implicit alignment argument from FixedHash

franzihei committed:
Update contributing.rst

Commit:
Improve phrasing.

Commit:
Update spelling and phrasing.

Commit:
Added functionality to semanticTest.cpp to ensure all output is printed on new lines. Also added a test (emit_multiple_outputs.sol).

Commit:
Added functionality to semanticTest.cpp to ensure all output is printed on new lines. Also added a test (emit_multiple_outputs.sol).

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,17 +508,27 @@ TestCase::TestResult SemanticTest::runTest(
 		{
 			ErrorReporter errorReporter;
-			_stream << test.format(
+			std::string outputString =
```
```suggestion
			string outputString =
```

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,17 +508,27 @@ TestCase::TestResult SemanticTest::runTest(
+			);
+
+			auto pos = outputString.find("  // ~");
+			while ((pos=outputString.find("  // ~", pos+6)) != std::string::npos) {
```
1. The formatting is a bit off - please use spaces between operators, i.e. `pos = outputString...` and `pos + 6`.
2. Have you maybe thought about using something from Boost? `boost::replace_all` seems to be a good candidate, although I haven't taken a detailed look.
3. A comment would be nice

cameel replied:
Also, `std::` here is superfluous. We use `using namespace std` in cpp files.

Andy53 replied:
I have added the additional spacing. This replaces all but the first instance of the pattern. I would assume boost::replace_all would replace all of them. I don't know of anything in boost that would do the same thing but if there is I am happy to change it.

nikola-matic commented on test/libsolidity/semanticTests/emit_multiple_outputs.sol:
```diff
@@ -0,0 +1,12 @@
+contract C {
+  event Terminated();
+
+  function terminate() external {
```
@cameel did we not have a single test that would emit two of the same events before (in a sequence)? I would have expected @Andy53's change to have altered some other tests.

Andy53 replied:
I didn't find another test that did this however there are a lot so I may have just missed it.

cameel replied:
Apparently we didn't, seeing that @axic reported the issue only 20 days ago.

That's probably because the support for testing emitted events is relatively new in isoltest (@aarlt implemented in ~1 year ago) and in older semantic tests emitting them was probably seen as not adding much to the test.

nikola-matic reviewed (commented):
Thanks for the PR Andy. Not everyone is back from DevCon, but we'll make a push to be a bit quicker with reviews.

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Andy53 committed:
Added additional spacing.

Andy53 committed:
Added additional spacing.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,17 +508,27 @@ TestCase::TestResult SemanticTest::runTest(
 				errorReporter,
 				_linePrefix,
 				m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
 				_formatted,
-				/* _interactivePrint */ true
-			) << endl;
+				true
```
Why remove the comment?

Andy53 replied:
I think the comment got deleted when I was messing with the method inputs. There is no reason I can think of to remove it so I will put it back in.

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,17 +508,27 @@ TestCase::TestResult SemanticTest::runTest(
 		{
 			ErrorReporter errorReporter;
-			_stream << test.format(
+			string outputString =
+			test.format(
```
```suggestion
			string outputString = test.format(
			    ...
			)
```

cameel commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,17 +508,27 @@ TestCase::TestResult SemanticTest::runTest(
+			auto pos = outputString.find("  // ~");
+			while ((pos=outputString.find("  // ~", pos+6)) != std::string::npos) {
+				outputString.replace(pos, 6, "\n  // ~");
+			}
```
This feels like a hack and has potential to cause some false-positives. I don't expect `  // ~` to happen a lot in the middle of expectations but it's also not *that* unlikely and hitting this corner case would be very confusing. Can you fix the issue at the source instead of patching the output?

Andy53 replied:
I have had another go at this and moved the line break to line 245 in SemanticTest::eventSideEffectHook. Let me know what you think.

I originally set it as a patch on the output so it would (hopefully) catch all instances of missing line breaks however I have moved it to the block that parses emits now.

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

Andy53 committed:
Moved line break to eventSideEffectHook

I moved the line break to SemanticTest::eventSideEffectHook line 245.

Andy53 committed:
Moved line break to eventSideEffectHook

I moved the line break to SemanticTest::eventSideEffectHook line 245.

Andy53 committed:
Moved endl;

Andy53 committed:
Moved endl;

Andy53 committed:
Updates

Andy53 committed:
Updates

Andy53 committed:
Finished (hopefully)

Andy53 committed:
Finished (hopefully)

Andy53 committed:
Added additional checks to fix failing tests.

Andy53 committed:
Added additional checks to fix failing tests.

Andy53 committed:
Update emit_multiple_outputs.sol

Andy53 committed:
Update emit_multiple_outputs.sol

Commit:
updates

Commit:
updates

Commit:
more updates

Commit:
added changes to testFunctionCall

Commit:
added changes to testFunctionCall

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -264,6 +264,19 @@ vector<string> SemanticTest::eventSideEffectHook(FunctionCall const&) const
 		sideEffects.emplace_back(sideEffect.str());
 	}
+	if(sideEffects.size() > 0){
+		for(long unsigned int i = 0; i < sideEffects.size() - 1; i++)
```
```suggestion
		for(size_t i = 0; i < sideEffects.size() - 1; i++)
```

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -264,6 +264,19 @@ vector<string> SemanticTest::eventSideEffectHook(FunctionCall const&) const
+	if(sideEffects.size() > 0){
+		for(long unsigned int i = 0; i < sideEffects.size() - 1; i++)
+		{
+			if(i == 0)
```
Wouldn't it make more sense to set this above the loop, since you're already checking that `sideEffects` is not empty, so the `sideEffects[0]` will always be present. This will allow you to avoid branching in the loop.

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -264,6 +264,19 @@ vector<string> SemanticTest::eventSideEffectHook(FunctionCall const&) const
 		sideEffect << joinHumanReadable(eventStrings);
 		sideEffects.emplace_back(sideEffect.str());
 	}
+	if(sideEffects.size() > 0){
```
```suggestion
	if(!sideEffects.empty()){
```

nikola-matic reviewed (changes requested):
Hey @Andy53, so I've taken a more detailed look at the PR, and it turns out that it's a fairly simple issue. Namely, the reason the expected results were printed on the same line is that in `TestFunctionCall::format` we were checking that the current side effect is not equal to the last side effect so as not to insert a new line after the last element is piped into the `stringstream`. 

In case of two identical emits, the first will be equal to the last, and a new line will therefore not be inserted. In order to solve this, the only necessary thing is to refactor the looping of elements in the `sideEffects` vector, i.e.:

```cpp
if (!sideEffects.empty())
{
	stream << std::endl;
	size_t i = 0;
	for (; i < sideEffects.size() - 1; ++i)
               stream << _linePrefix << "// ~ " << sideEffects[i] << std::endl;
	stream << _linePrefix << "// ~ " << sideEffects[i];
}
```

Nothing else is necessary here, so the changes in `SemanticTest.cpp` should be reverted.

nikola-matic commented:
> Hey @Andy53, so I've taken a more detailed look at the PR, and it turns out that it's a fairly simple issue. Namely, the reason the actual results were printed on the same line is that `TestFunctionCall::format` we were checking that the current side effect is not equal to the last side effect so as not to insert a new line after the last element is piped into the `stringstream`.
> 
> In case of two identical emits, the first will be equal to the last, and a new line will therefore not be inserted. In order to solve this, the only necessary thing is to refactor the looping of elements in the `sideEffects` vector, i.e.:
> 
> ```c++
> if (!sideEffects.empty())
> {
> 	stream << std::endl;
> 	size_t i = 0;
> 	for (; i < sideEffects.size() - 1; ++i)
> 	{
> 		stream << _linePrefix << "// ~ " << sideEffects[i] << std::endl;
> 	}
> 	stream << _linePrefix << "// ~ " << sideEffects[i];
> }
> ```
> 
> Nothing else is necessary here, so the changes in `SemanticTest.cpp` should be reverted.

Also, I've implemented this in your branch, so if you don't feel doing this yourself, I can pushed the changes :)

Commit:
updated TestFunctionCall::format

Commit:
updated TestFunctionCall::format

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -518,7 +520,8 @@ TestCase::TestResult SemanticTest::runTest(
 				/* _interactivePrint */ true
-			) << endl;
+			);
+
```
```suggestion
```

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,7 +508,9 @@ TestCase::TestResult SemanticTest::runTest(
 			_stream << errorReporter.format(_linePrefix, _formatted);
 		}
 		_stream << endl;
+
```
```suggestion
```

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -508,7 +508,9 @@ TestCase::TestResult SemanticTest::runTest(
 		_stream << endl;
+
 		AnsiColorized(_stream, _formatted, {BOLD, CYAN}) << _linePrefix << "Obtained result:" << endl;
+
```
```suggestion
```

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

updated TestFunctionCall::format

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Added additional test.

renamed emit tests

removed multiple_consecutive_emits.sol

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Andy53 committed:
parent 25611589922920fbb2413b94af249b1525e09980
author Andy <35195301+Andy53@users.noreply.github.com> 1666632594 -0600
committer andy53 <andy@evilrobots.club> 1666636112 -0600

parent 25611589922920fbb2413b94af249b1525e09980
author Andy <35195301+Andy53@users.noreply.github.com> 1666632594 -0600
committer andy53 <andy@evilrobots.club> 1666635990 -0600

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

updated TestFunctionCall::format

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Added additional test.

renamed emit tests

removed multiple_consecutive_emits.sol

Fix stripping in CLI tests does not produce correct JSON formatting

Pretty print expected json output of command line tests

Add v0.6.0 change notes for inline assembly opcode syntax changes.

Renaming lspAssert to lspRequire plus adding some documentation to make it more clear what this function is supposed to be used for.

EVMHost: Simplify some code and dcoument functions

Added functionality to semanticTest.cpp to ensure all output is printed on new lines. Also added a test (emit_multiple_outputs.sol).

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Added additional spacing.

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

Moved line break to eventSideEffectHook

I moved the line break to SemanticTest::eventSideEffectHook line 245.

Moved endl;

Updates

Finished (hopefully)

Added additional checks to fix failing tests.

Update emit_multiple_outputs.sol

updated TestFunctionCall::format

updated TestFunctionCall::format

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

updated TestFunctionCall::format

Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Added additional test.

renamed emit tests

removed multiple_consecutive_emits.sol

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -518,7 +518,7 @@ TestCase::TestResult SemanticTest::runTest(
 				m_gasCostFailure ? TestFunctionCall::RenderMode::ExpectedValuesActualGas : TestFunctionCall::RenderMode::ActualValuesExpectedGas,
 				_formatted,
 				/* _interactivePrint */ true
-			) << endl;
```
Can you add back the `endl` as well - `SemanticTest.cpp` should remain unchanged.

Andy53 replied:
I will get it put back now.

Commit:
updated TestFunctionCall::format

Commit:
updated TestFunctionCall::format

Commit:
updated TestFunctionCall::format

Commit:
updated TestFunctionCall::format

nikola-matic commented on test/libsolidity/SemanticTest.cpp:
```diff
@@ -519,6 +519,7 @@ TestCase::TestResult SemanticTest::runTest(
 				_formatted,
 				/* _interactivePrint */ true
 			) << endl;
+
```
```suggestion
```
Missed another one.

Andy53 committed:
Update test/libsolidity/SemanticTest.cpp

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>

Commit:
Added additional test.

Commit:
renamed emit tests

Commit:
removed multiple_consecutive_emits.sol

Commit:
removed white space from emit_two_identical_events.sol

Commit:
Merge branch 'render-multiple-emitted-logs-correctly' of github.com:Andy53/solidity into render-multiple-emitted-logs-correctly
